mqtt/mqtt-edgehub/src/connection/delivery.rs

```diff
@@ -0,0 +1,137 @@
+use std::{collections::HashMap, sync::Arc};
+
+use async_trait::async_trait;
+use lazy_static::lazy_static;
+use mqtt_broker::{
+    BrokerHandle, Error, IncomingPacketProcessor, Message, OutgoingPacketProcessor, PacketAction,
+    SystemEvent,
+};
+use parking_lot::Mutex;
+use regex::Regex;
+
+use mqtt3::proto::{self, Packet, PacketIdentifier, Publication};
+
+pub struct PublicationDelivery<P> {
+    inner: P,
+    broker_handle: BrokerHandle,
+    waited_to_be_acked: Arc<Mutex<HashMap<PacketIdentifier, String>>>,
+}
+
+/// MQTT packet processor wrapper. It identifies `IoTHub` M2M outgoing publishes,
+/// saves PACKETID to send a confirmation packet back to sender on special topic
+/// "$edgehub/delivery".
```

Comment:
I would also add "why" this is needed. That this is used as a back pressure mechanism to avoid M2M message loss.